concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -127,160 +191,123 @@ writeSection ::
-initialHandle :: FilePath -> IO Handle
-initialHandle p = do
-    createDirectoryIfMissing True (takeDirectory p)
-    openBinaryFile p WriteMode
-
--- |Export a database in V3 format, given the data directory root.
-exportDatabaseV3 ::
-    -- |Data directory
-    FilePath ->
-    -- |Export file name
-    FilePath ->
-    IO ()
-exportDatabaseV3 dbDir outFile = do
-    h <- initialHandle outFile
-    BS.hPut h $ encode (3 :: Version)
-    exportSections h dbDir 0
-    hClose h
-
--- |Export a series of blocks. When a block contains a finalization record, the 'dbsLastFinIndex'
--- field of the state is updated accordingly.
-exportBlocks ::
-    forall pv m.
+-- |Export a series of blocks as a chunk of a specified length.
+exportBlocksToChunk ::
+  forall pv m.
     (IsProtocolVersion pv, MonadIO m, MonadState (DBState pv) m, MonadCatch m) =>
     -- |Handle to export to
     Handle ->
     -- |Height of next block to export
     BlockHeight ->
-    m Word64
-exportBlocks hdl = eb 0
-  where
-    eb count height =
-        resizeOnResized (readFinalizedBlockAtHeight height) >>= \case
-            Nothing -> return count
+    -- |Number of blocks to export
+    Word64 ->
+    -- |(Block counter final value, finalization records not included in blocks)
+    m (Word64, [FinalizationIndex])
+exportBlocksToChunk hdl height count =
+  resizeOnResized (readFinalizedBlockAtHeight height) >>= \case
+            Nothing -> return (count, [])
             Just sb | NormalBlock normalBlock <- sbBlock sb -> do
                 let serializedBlock =
                         runPut $ putVersionedBlock (protocolVersion @pv) normalBlock
                     len = fromIntegral $ BS.length serializedBlock
                 liftIO $ do
                     BS.hPut hdl $ runPut $ putWord64be len
                     BS.hPut hdl serializedBlock
-                    putStrLn $
-                        "Exported block: " ++ show (_bpHash (sbInfo sb))
-                            ++ " at height "
-                            ++ show height
-                            ++ " with blockSlot "
-                            ++ show (blockSlot (sbBlock sb))
-                            ++ ". Serialized length: "
-                            ++ show (8 + len)
-                forM_ (blockFields (sbBlock sb)) $ \bf ->
-                    case blockFinalizationData bf of
-                        BlockFinalizationData fr ->
-                            dbsLastFinIndex .= finalizationIndex fr
-                        _ -> return ()
-                eb (count + 1) (height + 1)
-            Just _ -> return count -- this branch should never be reachable, genesis blocks are not exported.
+                let finRecIndex :: Maybe FinalizationIndex = join . forM (blockFields (sbBlock sb)) $ \bf ->
+                      case blockFinalizationData bf of
+                        BlockFinalizationData _ -> Nothing
+                        _ -> Just (sbFinalizationIndex sb)
```

Comment:
```suggestion
                let finRecIndex :: Maybe FinalizationIndex = (blockFields (sbBlock sb)) >>= \bf ->
                      case blockFinalizationData bf of
                        BlockFinalizationData _ -> Nothing
                        NoFinalizationData -> Just (sbFinalizationIndex sb)
```

Reply:
Please document the intent here. It is not at all obvious why we do this here, so this is not maintainable.